crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6474 @@
+                    }),
+                _ => Ok(false),
+            }
+        };
+        rustworkx_core::dag_algo::collect_runs(&self.dag, filter_fn)
+            .map(|node_iter| node_iter.map(|x| x.unwrap()))
+    }
+
+    /// Return a set of non-conditional runs of 1q "op" nodes.
+    pub fn collect_1q_runs(&self) -> Option<impl Iterator<Item = Vec<NodeIndex>> + '_> {
+        let filter_fn = move |node_index: NodeIndex| -> Result<bool, Infallible> {
+            let node = &self.dag[node_index];
+            match node {
+                NodeType::Operation(inst) => Ok(inst.op.num_qubits() == 1
+                    && inst.op.num_clbits() == 0
+                    && inst.op.matrix(inst.params_view()).is_some()
+                    && inst.condition().is_none()),
+                _ => Ok(false),
+            }
+        };
+        rustworkx_core::dag_algo::collect_runs(&self.dag, filter_fn)
+            .map(|node_iter| node_iter.map(|x| x.unwrap()))
+    }
+
+    /// Return a set of non-conditional runs of 2q "op" nodes.
+    pub fn collect_2q_runs(&self) -> Option<Vec<Vec<NodeIndex>>> {
+        let filter_fn = move |node_index: NodeIndex| -> Result<Option<bool>, Infallible> {
+            let node = &self.dag[node_index];
+            match node {
+                NodeType::Operation(inst) => match inst.op.view() {
+                    OperationRef::Standard(gate) => Ok(Some(
+                        gate.num_qubits() <= 2
+                            && inst.condition().is_none()
+                            && !inst.is_parameterized(),
+                    )),
+                    OperationRef::Gate(gate) => Ok(Some(
+                        gate.num_qubits() <= 2
+                            && inst.condition().is_none()
+                            && !inst.is_parameterized(),
+                    )),
+                    _ => Ok(Some(false)),
+                },
+                _ => Ok(None),
+            }
+        };
+
+        let color_fn = move |edge_index: EdgeIndex| -> Result<Option<usize>, Infallible> {
+            let wire = self.dag.edge_weight(edge_index).unwrap();
+            match wire {
+                Wire::Qubit(index) => Ok(Some(index.0 as usize)),
+                _ => Ok(None),
+            }
+        };
+        rustworkx_core::dag_algo::collect_bicolor_runs(&self.dag, filter_fn, color_fn).unwrap()
+    }
+
+    fn increment_op(&mut self, op: String) {
+        self.op_names
+            .entry(op)
+            .and_modify(|count| *count += 1)
+            .or_insert(1);
+    }
+
+    fn decrement_op(&mut self, op: String) {
+        match self.op_names.entry(op) {
+            Entry::Occupied(mut o) => {
+                if *o.get() > 1usize {
+                    *o.get_mut() -= 1;
+                } else {
+                    o.swap_remove();
+                }
+            }
+            _ => panic!("Cannot decrement something not added!"),
+        }
+    }
```

Comment:
Unfortunately because we need to use an `IndexMap` for reproducibility we can't use `entry_ref()` which would be my preference here. It seems to only be a thing on `hashbrown::HashMap`.

Reply:
Updated in: https://github.com/Qiskit/qiskit/pull/12550/commits/694d78e0dc5fc4fd5da01805ed6007c7e3b92346